### PR TITLE
Add support for custom panic handler

### PR DESCRIPTION
**Pull request by me-no-dev**
redo for: https://github.com/espressif/arduino-esp32/pull/8574

### CODE REVIEW

File: cores/esp32/esp32-hal-misc.c [hunk 1]
```diff
@@ -322,3 +322,93 @@ const char * ARDUINO_ISR_ATTR pathToFileName(const char * path)
...
+static void * _panic_handler_arg = NULL;
+
+void set_arduino_panic_handler(arduino_panic_handler_t handler, void * arg){
+    _panic_handler = handler;
+    _panic_handler_arg = arg;
+}
+
+arduino_panic_handler_t * get_arduino_panic_handler(void){
+    return _panic_handler;
+}
```
Comment (lucasssvaz) [inline cores/esp32/esp32-hal-misc.c:345]:
```suggestion
arduino_panic_handler_t get_arduino_panic_handler(void){
    return _panic_handler;
}
```

File: cores/esp32/esp32-hal.h [hunk 2]
```diff
@@ -137,6 +137,22 @@ void arduino_phy_init();
...
+    bool backtrace_corrupt;     // if backtrace is corrupt
+    bool backtrace_continues;   // if backtrace continues, but did not fit
+    unsigned int backtrace_len; // number of backtrace addresses
+    unsigned int backtrace[60]; // backtrace addresses array
+} arduino_panic_info_t;
+
+typedef void (*arduino_panic_handler_t)(arduino_panic_info_t * info, void * arg);
+
+void set_arduino_panic_handler(arduino_panic_handler_t handler, void * arg);
+arduino_panic_handler_t * get_arduino_panic_handler(void);
```
Comment (lucasssvaz) [inline cores/esp32/esp32-hal.h:153]:
```suggestion
arduino_panic_handler_t get_arduino_panic_handler(void);
```

**Review by lucasssvaz — CHANGES_REQUESTED** [review]
Change `get_arduino_panic_handler` to avoid compilation warning. `arduino_panic_handler_t` is already a pointer

**Review by lucasssvaz — APPROVED** [review]

**Comment by Laxilef** [comment]
Hi @me-no-dev 

Please explain how to work with the ``set_arduino_panic_handler``? Previously, I created an array of pairs and filled it like this:
```c++
std::make_pair(esp_cpu_process_stack_pc(stk_frame.pc), stk_frame.sp);
```
But now the backtrace array in the ``arduino_panic_info_t`` structure is filled only PC:
```c++
    pc_ptr = esp_cpu_process_stack_pc(stk_frame.pc);
    if (esp_ptr_executable((void *)pc_ptr)) {
      p_info.backtrace[p_info.backtrace_len++] = pc_ptr;
    // ...
```

How can I get SP? Thanks.

**Pull request merged**